ledger/execution/tx_coinbase.go

```diff
@@ -378,54 +381,34 @@ func grantEliteEdgeNodeReward(ledger core.Ledger, view *st.StoreView, guardianVo
 		return
 	}
 
-	totalStake := big.NewInt(0)
-	eliteEdgeNodePool = eliteEdgeNodePool.WithStake()
-	for i, e := range eliteEdgeNodePool.SortedEliteEdgeNodes {
-		if eliteEdgeNodeVotes.Multiplies[i] == 0 {
-			continue
-		}
-		totalStake.Add(totalStake, e.TotalStake())
-	}
-
-	if totalStake.Cmp(big.NewInt(0)) != 0 {
-		stakeSourceMap := map[common.Address]*big.Int{}
-		stakeSourceList := []common.Address{}
+	stakeSourceMap := map[common.Address]*big.Int{}
+	totalStake := new(big.Int)
 
-		for i, e := range eliteEdgeNodePool.SortedEliteEdgeNodes {
-			if eliteEdgeNodeVotes.Multiplies[i] == 0 {
+	for _, eenAddr := range eliteEdgeNodeVotes.Addresses {
+		weight := eliteEdgeNodePool.RandomRewardWeight(eliteEdgeNodeVotes.Block, eenAddr)
+		een := eliteEdgeNodePool.Get(eenAddr)
+		for _, stake := range een.Stakes {
+			if stake.Withdrawn {
 				continue
 			}
-			stakes := e.Stakes
-			for _, stake := range stakes {
-				if stake.Withdrawn {
-					continue
-				}
-				stakeAmount := stake.Amount
-				stakeSource := stake.Source
-				if stakeAmountSum, exists := stakeSourceMap[stakeSource]; exists {
-					stakeAmountSum.Add(stakeAmountSum, stakeAmount)
-				} else {
-					stakeSourceMap[stakeSource] = stakeAmount
-					stakeSourceList = append(stakeSourceList, stakeSource)
-				}
+			if _, ok := stakeSourceMap[stake.Source]; !ok {
+				stakeSourceMap[stake.Source] = new(big.Int)
 			}
+			weightedStake := new(big.Int).Mul(stake.Amount, big.NewInt(int64(weight)))
+			stakeSourceMap[stake.Source].Add(stakeSourceMap[stake.Source], weightedStake)
```

Comment:
Maybe we don't need to multiply the weight with stake.Amount, since the weight already encodes the relative stake? Reference:

https://github.com/thetatoken/theta-protocol-ledger/blob/4f004306fd1e2b768679ba84dd07262dd4eb565a/ledger/execution/tx_coinbase.go#L509